platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/problems/ConfigurationCacheProblemsSummary.kt

```diff
@@ -88,29 +89,41 @@ class ConfigurationCacheProblemsSummary(
      */
     fun onProblem(problem: PropertyProblem, severity: ProblemSeverity): Boolean {
         lock.withLock {
-            problemCount += 1
+            totalProblemCount += 1
             when (severity) {
-                ProblemSeverity.Failure -> failureCount += 1
-                ProblemSeverity.Suppressed -> suppressedCount += 1
-                ProblemSeverity.Warning -> {}
-                ProblemSeverity.Info -> {}
+                ProblemSeverity.Deferred -> deferredProblemCount += 1
+                ProblemSeverity.Suppressed -> suppressedProblemCount += 1
+                ProblemSeverity.Interrupting -> {}
             }
             if (overflowed) {
                 return false
             }
-            if (problemCount > maxCollectedProblems) {
+            if (totalProblemCount > maxCollectedProblems) {
                 overflowed = true
                 return false
             }
-            val uniqueProblem = UniquePropertyProblem.of(problem)
-            if (uniqueProblems.add(uniqueProblem) && causes.size < MAX_CAUSES) {
+            val isNew = trackUniqueProblems(problem, severity)
+            if (isNew && causes.size < MAX_CAUSES) {
                 problem.exception?.let {
                     causes.add(it)
                 }
             }
             return true
         }
     }
+
+    /**
+     * Return true if no similar problem has been seen before.
+     */
+    private
+    fun trackUniqueProblems(problem: PropertyProblem, severity: ProblemSeverity): Boolean {
+        val reducedProblem = UniquePropertyProblem.of(problem)
+        val isNew = !uniqueProblems.containsKey(reducedProblem)
+        uniqueProblems.merge(reducedProblem, severity) { old, new ->
+            if (severityComparator.compare(old, new) < 0) old else new
+        }
```

Comment:
💅 A _unique_ problem can be encountered more than once, with a different severity each time?  _unique_ sounds like the wrong word in this case.

Reply:
I generally agree that the naming should be updated here, but I'm reluctant to do it in this PR.

Though, different severities are not exactly related to this "uniqueness", as the severity is contextual. And we are collecting the most "relevant" severity, so they live together with the "unique" problems for convenience and memory saving reasons.

Reply:
I understand the intent after reading your explanation here, but the next time I read this code I imagine I'll forget and be back to thinking "wait, these aren't unique".

I don't understand the reluctance to rename this as part of this PR.  This class is already changing significantly, and this is private field with an increasingly misleading name.

Reply:
My reluctance was mainly based on the desire to save time and unblock follow-ups required for 9.0. However, since the rest of the reviews still take time, I returned to this and did the rename.

I renamed "unique problems" to "problem causes". I also added some javadocs to describe the intended purpose. 

See this last commit and let me know what you think:
- [`59fa4ae` (#33541)](https://github.com/gradle/gradle/pull/33541/commits/59fa4ae9b69f10f2b5ef1724190e42855e0231a1)